app/services/totp_creator.go

```diff
@@ -0,0 +1,48 @@
+package services
+
+import (
+	"bytes"
+	"image/png"
+
+	"github.com/keratin/authn-server/app/data"
+	"github.com/keratin/authn-server/lib/route"
+	"github.com/pkg/errors"
+	"github.com/pquerna/otp/totp"
+)
+
+// func TOTPCreator(
+// 	accountStore data.AccountStore, refreshTokenStore data.RefreshTokenStore, keyStore data.KeyStore, actives data.Actives, cfg *app.Config, reporter ops.ErrorReporter,
+// 	accountID int, audience *route.Domain, existingToken *models.RefreshToken,
+// ) (string, string, error) {
+
+//TOTPCreator handles the creation and storage of new TOTP tokens
+func TOTPCreator(accountStore data.AccountStore, totpCache data.TOTPCache, accountID int, audience *route.Domain) (*bytes.Buffer, error) {
+
+	account, err := AccountGetter(accountStore, accountID)
+	if err != nil {
+		return nil, err
+	}
+
+	//Generate totp key
+	key, err := totp.Generate(totp.GenerateOpts{
+		Issuer:      audience.Hostname,
+		AccountName: account.Username,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if err := totpCache.CacheTOTPSecret(account.ID, []byte(key.Secret())); err != nil {
+		return nil, errors.Wrap(err, "TOTPCreator")
+	}
```

Comment:
> crazy thought: do we need to keep a copy on the server during this process? the best argument i can see is that it allows us to expire the attempt. it really feels like i'm missing something obvious that my past self felt more confident about. 😬

The secret is plain text and shared between the client and the server. If AuthN doesn't keep the copy on the server then technically the client could ignore the secret AuthN sends and generate their own. This would allow the client to send whatever secret they want as long as they also send a code that verifies against it. Allowing clients the ability to choose the shared secret seems a little icky to me but I'm definitely not an expert. I can't see anything in [this document](https://tools.ietf.org/html/rfc6238) that would prevent the client from generating it if the AuthN secret was ignored.